# EDA Tutorial  


🏀 NBA dataset <br>

Surprisingly basketball is quite popular in Canada and among immigrants. <br><br>
I have never got into the game of basketball and all my basketball knowledge comes from a Japanese comic book called "Slum Dunk" and occucional sports news I come by.<br>

It will be interesting to see if the salary is related to the position, age, height or/and weight of a player. <br><br>
Do you think taller players get a higher salary or maybe certain positions are associated with a higher salary? <br><br>
It is my impression that taller players take a position under the nets and shorter players are more in the middle of the court moving the ball. <br>
<br>
Let's go find out with EDA. 

## 💠 Install and Imports

Install altair

`conda install python-graphviz` &emsp; /
`pip install graphviz` <br><br>
`conda install -c conda-forge altair vega_datasets` &emsp; /`pip install altair` 


In [1]:
import altair as alt
alt.themes.enable('fivethirtyeight')
import pandas as pd

### altair

✨ <a href="https://altair-viz.github.io/index.html">Altair</a>
 
 &emsp; [Check Example Gallery](https://altair-viz.github.io/gallery/index.html)

### pandas

✨ <a href="https://pandas.pydata.org/docs/user_guide/index.html">pandas</a>

pandas DataFrame 
        is a Two-Dimensional data structure, immutable, heterogeneous tabular data structure with labeled axes rows, and 
        columns. pandas Dataframe is consists of three components principal, data, rows, and columns.

* DataFrames support named rows & columns (you can also provides names to rows)
* Supports Hetrogenous Collections of data.
* DataFrame labeled axes (rows and columns).
* Can perform arithmetic operations on rows and columns.
* Supporting reading flat files like CSV,Excel, JSON and also reads SQL tables’s
* Handling of missing data.

## 💠 Loading in the data

In [2]:
bball_df = pd.read_csv('nba.csv')

bball_df.head(3)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN

Successfully loaded the original CSV file 

## 💠EDA

In [3]:
bball_df.shape

(458, 9)

There are 9 columns and 458 rows in this dataframe. 

### df.info()

Use `info()` to

✨ Check missing value <br>
⭐ Check datatype. <br> 
&emsp;&emsp; Any dates??

In [4]:
bball_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


There are 9 columns but only 457 rows.<br>
It is missing one row in every column. <br>
Let's check the last row.

In [5]:
bball_df.tail(5)

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

There is a mystery last row. <br>
It could be nothing or there might be a new player who is contracted but not revealed yet.<br>
There is not enough information to determine with the data given and my knowledge of NBA.<br>
Players without the`Name` will be removed from the dataframe.

✨ Missing values

&emsp; There is the last row which has NaN in all columns. <br>
&emsp; College and Salary columns are missing some values as well. 

⭐ Datatype

&emsp; The datatype for Number columns is float but the object might be the better choice.<br>
&emsp; The datatype for Height is an object but it needs to be `float` in order to be usable for analysis.  

Let's modify the original bball dataframe. <br>
The row without the Player name will be removed from nba_df.<br>
The datatype is also modified in the `Number` and `Height` Column. 

In [6]:
nba_df = bball_df.dropna(subset=['Name'])

Let's check that the last row is removed.

In [7]:
nba_df.tail()

Name       Team  Number Position   Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

In [8]:
nba_df.shape

(457, 9)

This indicates that there was only one row that was missing `Name` since only one row is removed.

In [9]:
nba_df['Number'] =nba_df['Number'].astype(int).astype(str)

C:\Users\sayo9\AppData\Local\Temp\ipykernel_21984\3738854278.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nba_df['Number'] =nba_df['Number'].astype(int).astype(str)


In [10]:
nba_df.head()

Name            Team Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics      0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics     99       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics     30       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics     28       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics      8       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

Now that the `Number` column represents the player number more accurately,<br>
Let's work on converting `Height` column into `cm` float value.

In [11]:
def convert_height(df):
    for i in range(0,df.shape[0]):
        raw_string = df.loc[i,['Height']]
        feet = raw_string.str.split('-')[0][0]
        inch = raw_string.str.split('-')[0][1]                 
        nba_df.loc[i,'Height'] = int(feet)*12*2.54 + int(inch)*2.54      
    return nba_df

In [12]:
nba_df =convert_height(nba_df)

In [13]:
nba_df['Height'] = nba_df['Height'].astype(float)

C:\Users\sayo9\AppData\Local\Temp\ipykernel_21984\800865302.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nba_df['Height'] = nba_df['Height'].astype(float)


Let's check the cleaned dataframe, `nba_df`. 

In [14]:
nba_df.head(5)

Name            Team Number Position   Age  Height  Weight  \
0  Avery Bradley  Boston Celtics      0       PG  25.0  187.96   180.0   
1    Jae Crowder  Boston Celtics     99       SF  25.0  198.12   235.0   
2   John Holland  Boston Celtics     30       SG  27.0  195.58   205.0   
3    R.J. Hunter  Boston Celtics     28       SG  22.0  195.58   185.0   
4  Jonas Jerebko  Boston Celtics      8       PF  29.0  208.28   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [15]:
nba_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    object 
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    float64
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 51.9+ KB


### df.describe()

Use `describe()` to get statistical data

In [16]:
nba_df.describe()

Age      Height      Weight        Salary
count  457.000000  457.000000  457.000000  4.460000e+02
mean    26.938731  201.143545  221.522976  4.842684e+06
std      4.404016    8.718404   26.368343  5.229238e+06
min     19.000000  175.260000  161.000000  3.088800e+04
25%     24.000000  195.580000  200.000000  1.044792e+06
50%     26.000000  203.200000  220.000000  2.839073e+06
75%     30.000000  208.280000  240.000000  6.500000e+06
max     40.000000  220.980000  307.000000  2.500000e+07

In [17]:
nba_df.describe(include='all')

Name                  Team Number Position         Age  \
count             457                   457    457      457  457.000000   
unique            457                    30     53        5         NaN   
top     Avery Bradley  New Orleans Pelicans      5       SG         NaN   
freq                1                    19     23      102         NaN   
mean              NaN                   NaN    NaN      NaN   26.938731   
std               NaN                   NaN    NaN      NaN    4.404016   
min               NaN                   NaN    NaN      NaN   19.000000   
25%               NaN                   NaN    NaN      NaN   24.000000   
50%               NaN                   NaN    NaN      NaN   26.000000   
75%               NaN                   NaN    NaN      NaN   30.000000   
max               NaN                   NaN    NaN      NaN   40.000000   

            Height      Weight   College        Salary  
count   457.000000  457.000000       373  4.460000e+02  
unique         NaN         NaN       118           NaN  
top            NaN         NaN  Kentucky           NaN  
freq           NaN         NaN        22           NaN  
mean    201.143545  221.522976       NaN  4.842684e+06  
std       8.718404   26.368343       NaN  5.229238e+06  
min     175.260000  161.000000       NaN  3.088800e+04  
25%     195.580000  200.000000       NaN  1.044792e+06  
50%     203.200000  220.000000       NaN  2.839073e+06  
75%     208.280000  240.000000       NaN  6.500000e+06  
max     220.980000  307.000000       NaN  2.500000e+07

In [18]:
nba_df.Position.unique()

array(['PG', 'SF', 'SG', 'PF', 'C'], dtype=object)

There are 5 positions. <br>
If you are not familiar with those positions,
here is <a href ='https://en.wikipedia.org/wiki/Basketball'>Wikipedia <br>


&emsp; PG - point guard <br>
&emsp; SF - shooting forward <br>
&emsp; SG - shooting guard <br>
&emsp; PF - power forward <br>
&emsp; C  - center

### .repeat()

In [19]:
numerical_columns = nba_df.select_dtypes('number').columns.tolist()
categorical_columns = nba_df.select_dtypes('object').columns.tolist()

In [20]:
alt.Chart(nba_df).mark_bar().encode(
     alt.X(alt.repeat(), type='quantitative', bin=alt.Bin(maxbins=25)),
     y='count()').properties(width=250, height=150).repeat(numerical_columns)

alt.RepeatChart(...)

It all shows very valuable distribution patterns.
<br>
Let's look at the shortest players.

In [21]:
nba_df.sort_values(by='Height').head()

Name              Team Number Position   Age  Height  Weight  \
11   Isaiah Thomas    Boston Celtics      4       PG  27.0  175.26   185.0   
130   Phil Pressey      Phoenix Suns     25       PG  25.0  180.34   175.0   
203      Ty Lawson    Indiana Pacers     10       PG  28.0  180.34   195.0   
22    Shane Larkin     Brooklyn Nets      0       PG  23.0  180.34   175.0   
228     J.J. Barea  Dallas Mavericks      5       PG  31.0  182.88   185.0   

            College     Salary  
11       Washington  6912869.0  
130        Missouri    55722.0  
203  North Carolina   211744.0  
22       Miami (FL)  1500000.0  
228    Northeastern  4290000.0

There is only one player who is under 180cm but he has made the top 25 % of the salary. <br>
This is very encouraging data for shorter players but in other words, it is extremely rare to make it to NBA if you are shorter than 180cm. 

In [22]:
alt.Chart(nba_df).mark_point(size=10).encode(
     alt.X(alt.repeat('column'), type='quantitative'),
     alt.Y(alt.repeat('row'), type='quantitative')
 ).properties(width=120, height=120
).repeat(column=numerical_columns, row=numerical_columns)

alt.RepeatChart(...)

Let's look at the relation of the columns. <br>
`Weight` and `Height` have postitive relation.<br>
As `Age` goes up, there are some players who make more salary and others who make less. <br>
`Height` and `Salary` do not seem to have a positive relation, but almost everybody is taller than 180cm. 

In [23]:
alt.Chart(nba_df).mark_bar().encode(
    alt.X('count()'),
    alt.Y(alt.repeat(), type='nominal', sort='x')
).properties(width=200, height=250).repeat(categorical_columns, columns =3)

alt.RepeatChart(...)

All teams have around 15 players (range: 13 - 19). <br>
Some numbers are more popular. <br>
Some positions have more players than others.<br>
Some colleges have produced a lot more NBA players. Here are the top 10 colleges. 

In [24]:
nba_df.groupby('College').size().sort_values(ascending=False).head(10)

College
Kentucky          22
Duke              20
Kansas            18
North Carolina    16
UCLA              15
Arizona           13
Florida           10
Texas              9
Syracuse           8
Connecticut        7
dtype: int64

### . facet()

In order to look closer into `Position` which has 5 unique values, `facet()` is useful.

In [25]:
alt.Chart(nba_df).mark_bar().encode(
    alt.X('Height', bin=alt.Bin(maxbins=30)),
    alt.Y('count()'),
    alt.Color('Position')
).properties(width=200, height=50
).facet('Position',columns=1)

alt.FacetChart(...)

In [26]:
alt.Chart(nba_df).mark_bar().encode(
    alt.X('Salary', bin=alt.Bin(maxbins=30)),
    alt.Y('count()'),
    alt.Color('Position')
).properties(width=200, height=140
).facet('Position',columns=3)

alt.FacetChart(...)

In [27]:
nba_df.sort_values(by='Salary', ascending=False).head()

Name                 Team Number Position   Age  Height  \
109      Kobe Bryant   Los Angeles Lakers     24       SF  37.0  198.12   
169     LeBron James  Cleveland Cavaliers     23       SF  31.0  203.20   
33   Carmelo Anthony      New York Knicks      7       SF  32.0  203.20   
251    Dwight Howard      Houston Rockets     12        C  30.0  210.82   
339       Chris Bosh           Miami Heat      1       PF  32.0  210.82   

     Weight       College      Salary  
109   212.0           NaN  25000000.0  
169   250.0           NaN  22970500.0  
33    240.0      Syracuse  22875000.0  
251   265.0           NaN  22359364.0  
339   235.0  Georgia Tech  22192730.0

The `Position` is associated with `Height` and Point Guard has the shortest players but the they all have similar pattern for the `Salary`.<br>
When you look at the top 5 highest salary players, it has three different positions. <br>
The person who made the highest salary from this NBA dataset is Kobe Bryant.